### PR TITLE
Support rope scaling

### PR DESCRIPTION
**Pull request by milocress**
Adds support for llama3 Rope scaling.

### CODE REVIEW

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -45,6 +45,7 @@
 import logging
 
 from transformers import PreTrainedModel, PreTrainedTokenizerBase
+from transformers.models.llama.modeling_llama import LlamaRotaryEmbedding
```
Comment (ShashankMosaicML) [inline llmfoundry/models/mpt/modeling_mpt.py:48]: I think this line is the same as line 57?

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 2]
```diff
@@ -88,14 +90,59 @@
 log = logging.getLogger(__name__)
 
 
+class InvalidConfigAccessError(KeyError):
+    pass
+
+
+_ALLOWED_LLAMA_CONFIG_KEYS = {
+    # these are the only config keys that are set and are safe to read from
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:98]:
```suggestion
    # These are the only config keys that are set and are safe to read from
```

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -88,14 +90,59 @@
...
+    'rope_scaling',
+    'rope_theta',
+    'max_position_embeddings',
+    'hidden_size',
+    'num_attention_heads',
+
+    # not set but llama modeling code tries to read this attribute
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:105]:
```suggestion
    # Not set but llama modeling code tries to read this attribute
```

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 4]
```diff
@@ -88,14 +90,59 @@
...
+    'partial_rotary_factor',
+
+    # benign transformers attributes needed for __init__
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:108]:
```suggestion
    # Benign transformers attributes needed for __init__
```

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 5]
```diff
@@ -88,14 +90,59 @@
...
+    '_get_generation_defaults',
+    'label2id',
+    'id2label',
+    'torch_dtype',
+    'problem_type',
+    '__class__',
+}
+
+
+class PartialLlamaConfig(LlamaConfig):
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:118]: Can you add a comment explaining why this class exists?

File: tests/models/test_rope_scaling.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    'rope_hf_config': {
+        'factor': 8.0,
+        'low_freq_factor': 1.0,
+        'high_freq_factor': 4.0,
+        'original_max_position_embeddings': 8192,
+        'type': 'llama3',
+    },
+}
+
+rope_dail_config = {}
```
Comment (dakinggg) [inline tests/models/test_rope_scaling.py:20]: delete?

**Review by dakinggg — COMMENTED** [review]
Note: when transformers 4.43.2 is released tomorrow, we should bump the min version to that in this PR

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 7]
```diff
@@ -134,6 +181,18 @@ def gen_rotary_embedding(
                 device=
                 'cpu',  # FSDP does not materialize modules with meta buffers, hence device is set to cpu
             )
+        elif rope_hf_config['type'] == 'llama3':
+            llama_rope_config = {**rope_hf_config}
+            llama_rope_config['rope_type'] = rope_hf_config.pop('type')
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:186]:
```suggestion
            llama_rope_config['rope_type'] = rope_hf_config.get('type')
```

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 8]
```diff
@@ -134,6 +181,18 @@ def gen_rotary_embedding(
                 device=
                 'cpu',  # FSDP does not materialize modules with meta buffers, hence device is set to cpu
```
Comment (snarayan21) [inline llmfoundry/models/mpt/modeling_mpt.py:182]: Can you also swap out the `dynamic`, `linear`, and `no_scaling` rope modifications to go through LlamaRotaryEmbedding and PartialLlamaConfig as well? transformers is deprecating the old code path

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 9]
```diff
@@ -134,6 +181,18 @@ def gen_rotary_embedding(
...
             )
+        elif rope_hf_config['type'] == 'llama3':
+            llama_rope_config = {**rope_hf_config}
+            llama_rope_config['rope_type'] = rope_hf_config.get('type')
+            return LlamaRotaryEmbedding(
+                config=PartialLlamaConfig(
```
Comment (snarayan21) [inline llmfoundry/models/mpt/modeling_mpt.py:188]: so i'm understanding -- the new class takes in just the complete llama config right

File: tests/models/test_rope_scaling.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+from transformers.models.llama.modeling_llama import LlamaRotaryEmbedding
```
Comment (snarayan21) [inline tests/models/test_rope_scaling.py:4]: can you add a test that checks that the new llama3 embedding type from HF outputs the same embeddings when `apply_scaling` is called as the rope modification code I shared with you?

**Review by snarayan21 — COMMENTED** [review]
some comments

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 11]
```diff
@@ -88,14 +90,59 @@
...
+
+        return super().__getattribute__(key)
+
+    def __getitem__(self, key: str):
+        if key not in _ALLOWED_LLAMA_CONFIG_KEYS:
+            raise InvalidConfigAccessError(key)
+
+        return super().__getitem__(key)
+
+    def _get_generation_defaults(self):
```
Comment (irenedea) [inline llmfoundry/models/mpt/modeling_mpt.py:132]: why do you need this? It always feels a little sketch to override private methods

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 5 above)
Comment (irenedea) [reply llmfoundry/models/mpt/modeling_mpt.py:118]: nit: docstring pls

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 9 above)
Comment (milocress) [reply llmfoundry/models/mpt/modeling_mpt.py:192]: yessir

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 11 above)
Comment (milocress) [reply llmfoundry/models/mpt/modeling_mpt.py]: I removed it the override it still works, so will delete.

File: tests/models/test_rope_scaling.py (hunk 6 above)
Comment (milocress) [reply tests/models/test_rope_scaling.py:20]: It's needed because the codepath is jank.

File: tests/models/test_rope_scaling.py (hunk 10 above)
Comment (dakinggg) [reply tests/models/test_rope_scaling.py:4]: i don't think that is necessary. we don't need to test _everything_ from hf

File: tests/models/test_rope_scaling.py (hunk 6 above)
Comment (dakinggg) [reply tests/models/test_rope_scaling.py:20]: oh right, i see

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 12]
```diff
@@ -108,32 +154,20 @@ def gen_rotary_embedding(
             'cpu',  # FSDP does not materialize modules with meta buffers, hence device is set to cpu
         )
     elif rope_impl == 'hf':
+        llama_rope_config = {**rope_hf_config}
+        llama_rope_config['rope_type'] = rope_hf_config.get('type')
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:158]:
```suggestion
        llama_rope_config['rope_type'] = rope_hf_config.get('type')
        if 'type' in llama_rope_config:
            del llama_rope_config['type']
```

Comment (dakinggg) [reply llmfoundry/models/mpt/modeling_mpt.py:158]: i think will fix the failing tests

**Review by dakinggg — APPROVED** [review]

**Pull request merged**